Design pattern creational factory

Bernard-McGeever opened the pull request:
### Description

I have added a short entry describing the design pattern; the factory. This includes a short definition, UML diagram (added to the media folder) and code example (written in Java.)

I wanted to write a series of short definitions on the creational patterns including:

Singleton
Builder
Prototype
**Factory**
Abstract Factory

This is to aid my own education.

### Type of Change

- Adding a new entry

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.

yangc95 commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,273 @@
+
+## UML Design
+
+![UML diagram of a factory](/media/factory-uml.png)
```
```suggestion
![UML diagram of a factory](https://raw.githubusercontent.com/Codecademy/docs/main/media/factory-uml.png)
```

Bernard-McGeever replied:
It might be an idea to put this URL in a readme. 

Apologies if I missed it

yangc95 commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,273 @@
+  }
+
+  public abstract void createTaxiRequest(final T pickUpLocation, final String destination, final int distance);
+
+
+  // getters and setters
```
```suggestion

  // Getters and setters
```

yangc95 commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
````diff
@@ -0,0 +1,273 @@
+}
+```
+
+**Note:** As the `pickUpLocation` property may be an address or a restaurant, we have made use of generics when modeling the `TaxiRequest`. The other property's object types are known.
````
```suggestion
**Note:** As the `pickUpLocation` property may be an address or a restaurant, generics are used when modeling the `TaxiRequest`. The other property's object types are known.
```

yangc95 commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,273 @@
+
+The constructor of `TaxiRequest` calls an abstract method `createTaxtRequest`. This enforces `TaxiRequest`'s subclasses to override the method with instantiation logic. Although the constructors of the subclasses may make use of the abstract super's constructor, instantiation logic remains in the concrete.
+
+Below provides the concrete subclasses of `StandardTaxtRequest`, `SevenSeaterTaxiRequest` and `MiniBusTaxiRequest` that all extend the abstract class above, `TaxiRequest`:
```
```suggestion
Below provides the concrete subclasses of `StandardTaxiRequest`, `SevenSeaterTaxiRequest`, and `MiniBusTaxiRequest` that all extend the abstract `TaxiRequest` class above:
```

yangc95 commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,273 @@
+
+The constructors for these concrete models, calls `super()` which intern calls the overridden method `createTaxiRequest`.
+
+The overridden method `createTaxtRequest()` handles the instantiation logic of the class. The properties `pickUpLocation`, `destination`, and`distance` are set, with the `estimatedPrice` programmatically calculated using the static property `FARE_MULTIPLIER`. A print statement has been added to help illustrated the pattern.
```
```suggestion
The overridden method `createTaxiRequest()` handles the instantiation logic of the class. The properties `pickUpLocation`, `destination`, and `distance` are set, with the `estimatedPrice` programmatically calculated using the static property `FARE_MULTIPLIER`. A print statement has been added to help illustrate the pattern.
```

yangc95 commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,273 @@
+
+  private final double price;
+
+  Restaurant(double price)
+  {
```
```suggestion
  Restaurant(double price) {
```

yangc95 commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,273 @@
+
+The factory method `getTaxiRequest()` has been overloaded using four different signatures.
+
+The first (`getTaxiRequest(String, String int)`) returns a `StandardTaxiRequest` object as the `pickUpLocation` is of type `String` and both `numberOfPassingers` and `TaxiType` have been negated from the method call.
+
+The second (`getTaxiRequest(Resteraunt, String, int)`) returns a `FoodDeliveryRequest` object as the `pickUpLocation` is of type `Restaurant`.
+
+The third (`getTaxiRequest(int, String, String, int)`) returns a taxi request based on the number of passengers provided and determined by a switch statement. Should the number of passengers exceed or fall below the required amount an exception is thrown.
+
+Finally, the fourth (`getTaxiRequest(TaxiType, String, String, int)`) returns a taxi request based on the `TaxiType` specified and determined by a switch statement. Should the `TaxiType` not be recognised, an exception is thrown.
```
```suggestion
The first `getTaxiRequest(String, String int)` that returns a `StandardTaxiRequest` object as the `pickUpLocation` is of type `String` and both `numberOfPassingers` and `TaxiType` have been negated from the method call.

The second `getTaxiRequest(Resteraunt, String, int)` that returns a `FoodDeliveryRequest` object as the `pickUpLocation` is of type `Restaurant`.

The third `getTaxiRequest(int, String, String, int)` that returns a taxi request is based on the number of passengers provided and is determined by a switch statement. Should the number of passengers exceed or fall below the required amount, an exception is thrown.

Finally, the fourth `getTaxiRequest(TaxiType, String, String, int)` that returns a taxi request based on the `TaxiType` is specified and determined by a switch statement. Should the `TaxiType` not be recognized, an exception is thrown.
```

yangc95 commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
````diff
@@ -0,0 +1,273 @@
+```java
+public enum Restaurant
+{
+  MACDONALD(5.30),
````
```suggestion
  MCDONALD(5.30),
```

yangc95 commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,273 @@
+    TaxiCreator.getTaxiRequest("1 Charming Av", "49 GoodFeeling Ln", 5);
+
+    System.out.print("Food Delivery (Restaurant) - ");
+    TaxiCreator.getTaxiRequest(Restaurant.MACDONALD,"49 GoodFeeling Ln", 5);
```
```suggestion
    TaxiCreator.getTaxiRequest(Restaurant.MCDONALD,"49 GoodFeeling Ln", 5);
```

yangc95 commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,273 @@
+---
+Title: 'What is the Factory Pattern?'
```
```suggestion
Title: 'Factory Pattern'
```

Structure change for Docs and present/future contributions. 🟡

yangc95 reviewed (commented):
👋  Review number 1 is underway

🏷️ Tag me once the appropriate changes are made! Thanks @Bernard-McGeever 🥇

Bernard-McGeever commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,273 @@
+
+**Note:** As the `pickUpLocation` property may be an address or a restaurant, we have made use of generics when modeling the `TaxiRequest`. The other property's object types are known.
+
+The constructor of `TaxiRequest` calls an abstract method `createTaxtRequest`. This enforces `TaxiRequest`'s subclasses to override the method with instantiation logic. Although the constructors of the subclasses may make use of the abstract super's constructor, instantiation logic remains in the concrete.
```
You cant remove 'in the' here as the meaning is different. We have a (in essence) 2 taxi request classes, one abstract and one concrete. The createTaxiRequest provides instantiation logic in the concrete class, and this is enforced by the same method name in the abstract class. Removing the 'in the' would mean that the method's logic is concrete. What we are saying is the method's logic lives in the concrete class (and not the abstract.)

yangc95 replied:
Could we then just append the word "class" so it reads "instantiation logic remains in the concrete class."

Thanks for the clarification! It makes much more sense that the method's logic is in the concrete class rather than the method's logic being concrete.

Bernard-McGeever commented:
@yangc95 I have gone through and addressed your comments. All okay with but one (due to a change in meaning.) I have resolved all and commented where needed.

Although the e.g. is quite large in this one, I hope it was clear enough, and helped explain some of the considerations one might face when implementing the pattern.

(Just a warning: The last one (abstract factory) is kind of two factories put on top of each other. I'm currently trying to work out how much code is suitable)

yangc95 commented:
> Although the e.g. is quite large in this one, I hope it was clear enough, and helped explain some of the considerations one might face when implementing the pattern.
> 
> (Just a warning: The last one (abstract factory) is kind of two factories put on top of each other. I'm currently trying to work out how much code is suitable)

Thanks for taking a deeper look into a fundamental part of this pattern. Definitely make a comment note if you decide to make any changes to the code throughout the review process to keep us in the loop. 🙂

SSwiniarski commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,265 @@
+  public abstract void createTaxiRequest(final T pickUpLocation, final String destination, final int distance);
+
+
+  // Getters and setters
```
Some dots or ellipsis here to indicate omitted code.

Bernard-McGeever replied:
The only code that is missing here is the getters and setters indicated by the comment. The createTaxiRequest() is stubbed out as its an abstract method. This is just to enforce the sub classes to have a method with the same signature. 

I have resolved, but if I mis understood, please feel free to re-open

SSwiniarski commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
````diff
@@ -0,0 +1,265 @@
+}
+```
+
+**Note:** As the `pickUpLocation` property may be an address or a restaurant, generics are used when modeling the `TaxiRequest`. The other property's object types are known.
````
```suggestion
**Note:** As the `pickUpLocation` property may be an address or a restaurant, generics are used when modeling the `TaxiRequest`. The other properties' object types are known.
```

SSwiniarski commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,265 @@
+**Note:** As the `pickUpLocation` property may be an address or a restaurant, generics are used when modeling the `TaxiRequest`. The other property's object types are known.
+
+The constructor of `TaxiRequest` calls an abstract method `createTaxiRequest`. This enforces `TaxiRequest`'s subclasses to override the method 
+with instantiation logic. Although the constructors of the subclasses may make use of the abstract super's constructor, instantiation logic remains in the concrete.
```
```suggestion
with instantiation logic. Although the constructors of the subclasses may make use of the abstract super's constructor, instantiation logic remains in the concrete subclass.
```

SSwiniarski commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
````diff
@@ -0,0 +1,265 @@
+}
+```
+
+As these model classes inherit their properties from the super, they each automatically have `pickUpLocation`, `destination`, and `distance` as fields. The `pickUpLocation` in these cases is an address, and therefore, `pickUpLocation` is of type `String`.
````
```suggestion
As these model classes inherit their properties from the base class, they each automatically have `pickUpLocation`, `destination`, and `distance` as fields. The `pickUpLocation` in these cases is an address, and therefore, `pickUpLocation` is of type `String`.
```

SSwiniarski commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,265 @@
+
+As these model classes inherit their properties from the super, they each automatically have `pickUpLocation`, `destination`, and `distance` as fields. The `pickUpLocation` in these cases is an address, and therefore, `pickUpLocation` is of type `String`.
+
+The constructors for these concrete models, calls `super()` which intern calls the overridden method `createTaxiRequest`.
```
```suggestion
The constructors for these concrete models, calls `super()` which in turn calls the overridden method `createTaxiRequest`.
```

SSwiniarski commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
````diff
@@ -0,0 +1,265 @@
+}
+```
+
+**Note:** When the restaurant type is chosen, the constructor provides the property `price` with the correct value.
````
```suggestion
*Note:* When the restaurant type is chosen, the constructor provides the property `price` with the correct value.
```

SSwiniarski commented on content/general/concepts/what-is-the-factory-pattern/what-is-the-factory-pattern.md:
```diff
@@ -0,0 +1,265 @@
+The first `getTaxiRequest(String, String int)` that returns a `StandardTaxiRequest` object as the `pickUpLocation` is of type `String` and both 
+`numberOfPassingers` and `TaxiType` have been negated from the method call.
+
+The second `getTaxiRequest(Resteraunt, String, int)` that returns a `FoodDeliveryRequest` object as the `pickUpLocation` is of type `Restaurant`.
```
```suggestion
The second `getTaxiRequest(Restaurant, String, int)` that returns a `FoodDeliveryRequest` object as the `pickUpLocation` is of type `Restaurant`.
```

SSwiniarski reviewed (dismissed):
@Bernard-McGeever, I've completed a review with a few comments. Let us know when you've had them addressed.

sonnynomnom commented:
Hi @Bernard-McGeever, once you make the changes, then I can merge it in!

Also, can you update all the files that you created recently to just **name** instead of having folder names and file names of **what-is-name**?

Bernard-McGeever commented:
Hi @sonnynomnom and @SSwiniarski,

Your comments have been addressed. Thanks for the review :)

sonnynomnom commented:
Made some more changes: https://github.com/Codecademy/docs/blob/main/content/general/concepts/factory-pattern/factory-pattern.md

sonnynomnom commented:
Updated it a little more: https://github.com/Codecademy/docs/commit/b5d453170a9ba92c4eadac140e7e5571fa038ffb
